packages/@react-native-windows/telemetry/src/telemetry.ts

```diff
@@ -4,207 +4,411 @@
+  'react-native-windows-init',
+];
+
+export const NuGetPackagesWeTrack: string[] = [
+  'Microsoft.UI.Xaml',
+  'Microsoft.Windows.CppWinRT',
+  'Microsoft.WinUI',
+];
+
+/**
+ * The Telemetry class is responsible for reporting telemetry for RNW CLI.
+ */
 export class Telemetry {
-  static client?: appInsights.TelemetryClient | undefined = undefined;
+  protected static client?: appInsights.TelemetryClient = undefined;
+  protected static options: TelemetryOptions = {
+    setupString: DefaultSetupString,
+    preserveErrorMessages: false,
+  };
 
-  static disable() {
+  protected static isTest: boolean = basePropUtils.isCliTest();
+  protected static commandInfo: CommandInfo = {};
+  protected static versionsProp: Record<string, string> = {};
+  protected static projectProp?:
+    | projectUtils.AppProjectInfo
+    | projectUtils.DependencyProjectInfo = undefined;
+
+  protected static reset(): void {
+    // Reset client
     if (Telemetry.client) {
-      Telemetry.client.config.disableAppInsights = true;
+      Telemetry.client.flush();
+      Telemetry.client = undefined;
     }
-    Telemetry.shouldDisable = true;
+
+    // Reset local members
+    Telemetry.options = {
+      setupString: DefaultSetupString,
+      preserveErrorMessages: false,
+    };
+    Telemetry.commandInfo = {};
+    Telemetry.versionsProp = {};
+    Telemetry.projectProp = undefined;
+  }
+
+  static isEnabled(): boolean {
+    return Telemetry.client !== undefined;
   }
 
-  static setup(preserveMessages?: boolean) {
-    if (Telemetry.isCI()) {
-      this.disable();
+  static getSessionId(): string {
+    return basePropUtils.getSessionId();
+  }
+
+  /** Sets up the Telemetry static to be used elsewhere. */
+  static async setup(options?: Partial<TelemetryOptions>) {
+    if (Telemetry.client) {
+      // Bail since we've already setup
       return;
     }
-    if (Telemetry.client) {
+
+    // Bail if we're in CI and not capturing CI
+    if (!this.isTest && basePropUtils.isCI() && !basePropUtils.captureCI()) {
       return;
     }
-    if (!process.env.RNW_CLI_TEST) {
-      appInsights.Configuration.setInternalLogging(false, false);
+
+    // Save off options for later
+    Object.assign(Telemetry.options, options);
+
+    Telemetry.setupClient();
+
+    await Telemetry.setupBaseProperties();
+
+    Telemetry.setupTelemetryProcessors();
+  }
+
+  /** Sets up Telemetry.client. */
+  private static setupClient() {
+    appInsights.Configuration.setInternalLogging(
+      Telemetry.isTest,
+      Telemetry.isTest,
+    );
+
+    Telemetry.client = new appInsights.TelemetryClient(
+      Telemetry.options.setupString,
+    );
+
+    // Uncomment for Fiddler testing
+    // Telemetry.client.config.proxyHttpUrl = 'http://localhost:8888';
+    // Telemetry.client.config.proxyHttpsUrl = 'http://localhost:8888';
+
+    Telemetry.client.config.disableAppInsights = Telemetry.isTest;
+    Telemetry.client.channel.setUseDiskRetryCaching(!Telemetry.isTest);
```

Comment:
Why enable only on non-test?

Reply:
We're don't want telemetry from running unit tests to actually get transmitted. This caching is a mechanism where-by, if a transmit fails, the lib saves it to disk and the next time someone tries to send telemetry with this lib, it'll resend it. We don't want test events saved to disk to be picked up later and transmitted when we're not running tests.